Sources/SwiftParser/Statements.swift

```diff
@@ -177,11 +178,22 @@ extension Parser {
           arena: self.arena
         )
       )
-    } while keepGoing != nil && self.hasProgressed(&loopProgress)
+    } while keepGoing != nil && !atConditionListTerminator(isGuardStatement: isGuardStatement) && self.hasProgressed(&loopProgress)
 
     return RawConditionElementListSyntax(elements: elements, arena: self.arena)
   }
 
+  mutating func atConditionListTerminator(isGuardStatement: Bool) -> Bool {
+    guard experimentalFeatures.contains(.trailingComma) else {
+      return false
+    }
+    // condition terminator is `else` for `guard` statements and start of statement body for `if` or `while` statements
+    if isGuardStatement {
+      return self.at(.keyword(.else))
+    }
```

Comment:
Missing `else` is a very common mistakes for `guard` statement. Probably parsing it as the body emit better diagnostics if it's looks like a body.
```swift
    if isGuardStatement {
      if self.at(.keyword(.else)) {
        return true
      }
      // Missing `else` is a common mistake, fall back to  lookahead `{` check.
    }
```